extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java

```diff
@@ -340,17 +343,53 @@ void beanConfigClasses(
                                 new AnnotationValue[] { createStringValue("prefix", configClass.getPrefix()) }));
             }
 
+            collectTypes(combinedIndex, configClass);
+
             beanConfigurationRegistry.produce(new BeanConfiguratorBuildItem(
                     beanRegistrationPhase.getContext()
                             .configure(configClass.getConfigClass())
-                            .types(configClass.getConfigClass())
-                            .qualifiers(qualifiers.toArray(new AnnotationInstance[] {}))
+                            .types(collectTypes(combinedIndex, configClass))
+                            .qualifiers(qualifiers.toArray(EMPTY_ANNOTATION_INSTANCES))
                             .creator(ConfigMappingCreator.class)
                             .param("type", configClass.getConfigClass())
                             .param("prefix", configClass.getPrefix())));
         }
     }
 
+    private Type[] collectTypes(CombinedIndexBuildItem combinedIndex, ConfigClassBuildItem configClass) {
+        IndexView index = combinedIndex.getIndex();
+        DotName configIfaceName = DotName.createSimple(configClass.getConfigClass().getName());
+        ClassInfo configIfaceInfo = index.getClassByName(configIfaceName);
+        if ((configIfaceInfo == null) || configIfaceInfo.interfaceNames().isEmpty()) {
+            return new Type[] { Type.create(configIfaceName, Kind.CLASS) };
+        }
+
+        Set<DotName> allIfaces = new HashSet<>();
+        allIfaces.add(configIfaceName);
+        collectInterfacesRec(configIfaceInfo, index, allIfaces);
+        Type[] result = new Type[allIfaces.size()];
+        int i = 0;
+        for (DotName iface : allIfaces) {
+            result[i++] = Type.create(iface, Kind.CLASS);
+        }
+        return result;
+    }
+
+    private static void collectInterfacesRec(ClassInfo current, IndexView index, Set<DotName> result) {
+        List<DotName> interfaces = current.interfaceNames();
+        if (interfaces.isEmpty()) {
```

Comment:
To avoid a (possible?) infinite recursion: 

```suggestion
        if (interfaces.isEmpty() || result.contains(current.name())) {
```